### PR TITLE
Feat nipt ftp

### PR DESCRIPTION
**Pull request by barrystokman**
## Description
*This PR adds/fixes uploading NIPT resulst to the CK ftp server*

### How to test
TC1: upload specific case:
- [x] `cg upload nipt case tendercoral --dry-run`

TC2: upload all available cases:
- [x] `cg upload nipt all --dry-run`

### Expected test outcome
TC1:
- [x] results file for `tendercoral` is found in housekeeper, on hasta and is uploaded by running `lftp sftp://SciLife:LifeSci@dmz.edi.sll.se -e "cd SciLife_Till_StarLims; put /home/proj/stage/housekeeper-bundles/tendercoral/2021-05-06/summary.csv; bye"`
- [x] Take a screenshot and attach or copy/paste the output.
TC2:
- [x] all eligible results files (here: `tendercoral`) are found in housekeeper, on hasta and are uploaded by running `lftp sftp://SciLife:LifeSci@dmz.edi.sll.se -e "cd SciLife_Till_StarLims; put /path/to/summary.csv; bye"`
- [x] 
- [x] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [x] tests executed by @barrystokman 
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [x] **MINOR** - when you add functionality in a backwards compatible manner

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch
- [ ] Inform to ...

### CODE REVIEW

**Comment by barrystokman** [comment]
CLI help text:
image
image
image
image

`cg upload nipt case tendercoral --dry-run`
image

`lftp sftp://SciLife:LifeSci@dmz.edi.sll.se -e "cd SciLife_Till_StarLims; put /home/proj/stage/housekeeper-bundles/tendercoral/2021-05-06/summary.csv; bye"`

`cg upload nipt all --dry-run`
image
image

File: cg/meta/upload/nipt/nipt.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+from cg.models.cg_config import CGConfig
+from cg.store import Store, models
+from cg.utils import Process
+from housekeeper.store import Store
+
+LOG = logging.getLogger(__name__)
+
+
+class NiptUploadAPI:
+    """API for NIPT upload via ftp"""
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:21]: I think it should be more clear to what we are uploading

File: cg/meta/upload/nipt/nipt.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        """Set dry run"""
+
+        LOG.info("Set dry run to %s", dry_run)
+        self.dry_run = dry_run
+
+    def get_housekeeper_results_file(self, case_id: str, tags: Optional[list] = None) -> str:
+        """Get the result file for a NIPT analysis from Housekeeper"""
+
+        if not tags:
+            tags = self.RESULT_FILE_TAGS
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:45]:
```suggestion
            tags: List[str] = self.RESULT_FILE_TAGS
```

File: cg/meta/upload/nipt/nipt.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+        hk_all_results_files = self.housekeeper_api.get_files(bundle=case_id, tags=tags)
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:47]:
```suggestion
        hk_all_results_files: List[hk_models.File] = self.housekeeper_api.get_files(bundle=case_id, tags=tags)
```
Perhaps not correct but please type hint

File: cg/meta/upload/nipt/nipt.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+        if not list(hk_all_results_files):
+            raise HousekeeperFileMissingError(
+                f"No results files found in Housekeeper for NIPT case {case_id}."
+            )
+
+        return hk_all_results_files.first().path
+
+    def get_results_file_path(self, hk_results_file) -> Path:
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:56]: Type hint
```suggestion
    def get_results_file_path(self, hk_results_file: str) -> Path:
```

File: cg/meta/upload/nipt/nipt.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        """Get the full path to the results file on Hasta"""
+
+        results_file = self.root_dir / Path(hk_results_file)
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:59]:
```suggestion
        results_file: Path = self.root_dir / hk_results_file
```
If root dir is a `Path` the other objects can be strings (or Paths)

File: cg/meta/upload/nipt/nipt.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    def get_all_upload_analyses(self) -> QueryModel:
+        """Gets all nipt analyses that are ready to be uploaded"""
+        latest_nipt_analyses = self.status_db.latest_analyses().filter(
+            models.Analysis.pipeline == Pipeline.FLUFFY
+        )
+
+        return latest_nipt_analyses.filter(models.Analysis.uploaded_at.is_(None))
+
+    def upload(self, results_file: Path) -> None:
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:74]: I think this should be named `upload_to_...` as we probably will use this API to upload to NIPTool as well

File: cg/meta/upload/nipt/nipt.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        """Get the full path to the results file on Hasta"""
+
+        results_file = self.root_dir / Path(hk_results_file)
+
+        if not results_file.exists():
+            raise FileNotFoundError(f"Results file {results_file} not found on hasta!")
+
+        return results_file
+
+    def get_all_upload_analyses(self) -> QueryModel:
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:66]: If possible please return something that exposes more what it holds like
```suggestion
    def get_all_upload_analyses(self) -> List[models.Analysis]:
```

File: cg/meta/upload/nipt/nipt.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        """Gets all nipt analyses that are ready to be uploaded"""
+        latest_nipt_analyses = self.status_db.latest_analyses().filter(
+            models.Analysis.pipeline == Pipeline.FLUFFY
+        )
+
+        return latest_nipt_analyses.filter(models.Analysis.uploaded_at.is_(None))
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:72]: Then it needs to be changed here as well, perhaps something like

```suggestion
        return [analysis for analysis in latest_nipt_analyses.filter(models.Analysis.uploaded_at.is_(None))]
```

File: cg/meta/upload/nipt/nipt.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        self.process.run_command(parameters=parameters, dry_run=self.dry_run)
+
+        if self.process.stderr:
+            LOG.info(f"NIPT Upload stderr:\n{self.process.stderr}")
+        if self.process.stdout:
+            LOG.info(f"NIPT Upload stdout:\n{self.process.stdout}")
+
+    def update_analysis_uploaded_at_date(self, case_id: str) -> models.Analysis:
+        """Updates analysis_uploaded_at for the uploaded analysis"""
+        case_obj = self.status_db.family(case_id)
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:92]:
```suggestion
        case_obj: models.Family = self.status_db.family(case_id)
```

File: cg/meta/upload/nipt/nipt.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        analysis_obj = case_obj.analyses[0]
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:93]:
```suggestion
        analysis_obj: models.Analysis = case_obj.analyses[0]
```

File: cg/meta/upload/nipt/nipt.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        analysis_obj.uploaded_at = dt.datetime.now()
+
+        return analysis_obj
+
+    def update_analysis_upload_started_date(self, case_id: str) -> models.Analysis:
+        """Updates analysis_upload_started_at for the uploaded analysis"""
+
+        case_obj = self.status_db.family(case_id)
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:101]:
```suggestion
        case_obj: models.Family = self.status_db.family(case_id)
```

File: cg/meta/upload/nipt/nipt.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        analysis_obj = case_obj.analyses[0]
```
Comment (moonso) [inline cg/meta/upload/nipt/nipt.py:102]:
```suggestion
        analysis_obj: models.Analysis = case_obj.analyses[0]
```

File: cg/models/cg_config.py [hunk 13]
```diff
@@ -63,8 +64,15 @@ class CommonAppConfig(BaseModel):
     deploy_config: Optional[str]
 
 
+class FluffyUploadConfig(BaseModel):
```
Comment (moonso) [inline cg/models/cg_config.py:67]: 👍

File: cg/cli/upload/nipt.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        results_file = nipt_upload_api.get_results_file_path(hk_results_file)
+        LOG.info(f"Results file found: {results_file}")
+        LOG.info(f"Starting upload!")
+        nipt_upload_api.update_analysis_upload_started_date(case_id)
+        nipt_upload_api.upload(results_file)
+        LOG.info(f"Upload finished!")
+        nipt_upload_api.update_analysis_uploaded_at_date(case_id)
+        if not dry_run:
+            status_db.commit()
+    except Exception as error:
```
Comment (moonso) [inline cg/cli/upload/nipt.py:44]: type hints!

**Review by moonso — CHANGES_REQUESTED** [review]
Great job @barrystokman , really nice code!
I had some comments

File: cg/meta/upload/nipt/nipt.py (hunk 7 above)
Comment (barrystokman) [reply cg/meta/upload/nipt/nipt.py:66]: What do you think of using `Iterable[models.Analysis]` instead? A `QueryModel` qualifies as an `Iterable` as far as `typing` is concerned, or so Pycharm tells me. Then we can skip the list comprehension in the return too.

File: cg/meta/upload/nipt/nipt.py (hunk 3 above)
Comment (barrystokman) [reply cg/meta/upload/nipt/nipt.py:47]: Turning this into `Iterable[hk_models.File]`, as suggested by Pycharm.

File: cg/meta/upload/nipt/nipt.py (hunk 1 above)
Comment (barrystokman) [reply cg/meta/upload/nipt/nipt.py]: fixed in d1e79180d6dac13d8d76a5a28085b6907a7ca4dc

File: cg/meta/upload/nipt/nipt.py (hunk 3 above)
Comment (barrystokman) [reply cg/meta/upload/nipt/nipt.py]: suggested fix in d1e79180d6dac13d8d76a5a28085b6907a7ca4dc

File: cg/meta/upload/nipt/nipt.py (hunk 6 above)
Comment (barrystokman) [reply cg/meta/upload/nipt/nipt.py]: suggested fix in d1e79180d6dac13d8d76a5a28085b6907a7ca4dc

File: cg/cli/upload/nipt.py (hunk 14 above)
Comment (barrystokman) [reply cg/cli/upload/nipt.py]: fixed d1e79180d6dac13d8d76a5a28085b6907a7ca4dc

**Review by moonso — APPROVED** [review]
Looks great @barrystokman , nice job!!

File: cg/meta/upload/nipt/nipt.py (hunk 1 above)
Comment (moonso) [reply cg/meta/upload/nipt/nipt.py]: I think this is better. Then we can elaborate in next PR when we add upload-to-NIPTool-functionality

**Comment by barrystokman** [comment]
Bumped:
image
Deployed:

**Pull request merged**